sys/build.rs

```diff
@@ -129,21 +127,57 @@ fn feature_to_define(name: impl AsRef<str>) -> String {
 }
 
 fn patch<D: AsRef<Path>, P: AsRef<Path>>(out_dir: D, patch: P) {
-    let mut child = Command::new("patch")
-        .arg("-p1")
-        .stdin(Stdio::piped())
-        .current_dir(out_dir)
-        .spawn()
-        .expect("Unable to execute patch, you may need to install it: {}");
+    struct Patches<'a>(&'a str);
+
+    // A backtracking attempt to find a valid diff when there are multiple patches in one file (also known as patchset)
+    impl<'a> Iterator for Patches<'a> {
+        type Item = Patch<'a, str>;
+        fn next(&mut self) -> Option<Self::Item> {
+            let mut range = self.0.len();
+            loop {
+                match Patch::from_str(&self.0[..range]) {
+                    Ok(x) if x.hunks().is_empty() => break None,
+                    Err(_) if range < 1 => break None,
+                    Ok(x) => {
+                        self.0 = &self.0[range..];
+                        break Some(x);
+                    }
+                    Err(_) => range -= 1,
+                }
+            }
+        }
+    }
+
     println!("Appliyng patch {}", patch.as_ref().display());
     {
-        let patch = fs::read(patch).expect("Unable to read patch");
-
-        let stdin = child.stdin.as_mut().unwrap();
-        stdin.write_all(&patch).expect("Unable to apply patch");
+        let out_dir = out_dir.as_ref();
+        let patch = fs::read_to_string(patch).expect("Unable to read patch");
+        let patch = dos2unix(&patch);
```

Comment:
Does we actually need that for both patch and source?
Does we need it at all?
I mean we may simply fix our patches if you have problem with it.

Reply:
How do you determine which file to patch?

Reply:
@stevefan1999-personal Course, I about line-ending conversion.

Reply:
> @stevefan1999-personal Course, I about line-ending conversion.

I'm not sure what you mean

Reply:
Why we need the following:
```rust
let patch = dos2unix(&patch);
```
and
```rust
let original = dos2unix(&original);
```

Reply:
I have CRLF in Windows and diffy does not support CRLF yet. There is no allocation cost if you have LF.

Reply:
I know this could turn political quickly, but since QuickJS is written with GCC in mind, and is written by Fabrice Bellard, a legendary open source developer, I'm pretty sure he wrote this while only running on Linux, and Linux is LF by default. 

Yet on Git and Windows, you have automatic LR to CRLF conversion. It is up to the user to turn it off, but it seems like it is a viral option, and I believe this would cause even more unwanted trouble to end-users. 

I'm not aware if `autocrlf` can be turned off per Git repository or can be disabled for Git submodules but anyway, it is almost cost-free if you run this on Unix-like operating system. We should use do this conversion to "normalize" anyway.

Reply:
@stevefan1999-personal 

Using this *.gitattributes* will, by default, force text files to use LF EOLs, with CRLF exceptions where required for WinOS or MSVC.

```
# .gitattributes (git attributes config file)

# default; use LF EOLs for text files
* text=auto eol=lf

# CRLF required; force required CRLF EOLs for WinOS BAT/CMD and MSVC SLN/VCPROJ/VCXPROJ files
*.[bB][aA][tT]                      text eol=crlf
*.[cC][mM][dD]                      text eol=crlf
*.[sS][lL][nN]                      text eol=crlf
*.[Vv][Cc][Pp][Rr][Oo][Jj]          text eol=crlf
*.[Vv][Cc][Pp][Rr][Oo][Jj].*        text eol=crlf
*.[Vv][Cc][Xx][Pp][Rr][Oo][Jj]      text eol=crlf
*.[Vv][Cc][Xx][Pp][Rr][Oo][Jj].*    text eol=crlf
```